### PR TITLE
[WIP] - Multicam Display Support

### LINKED ISSUE #1094: Multi-Cam Simultaneous Display [FEATURE]
Having multiple cameras side-by-side in the main viewing area instead of utilizing the drop-down tool.

Tasks:
- Main geoJS context needs to be loaded multiple times
- Includes multiple loading of Image/Video annotators, each Media Annotator has it's own source video/images.
- Syncing of Media Annotators, current time/frame synchronization
- A LayerManager for each camera connected to the proper dataset in the trackStore.  So each camera view will draw it's own annotations in it's own GeoJS context/layers.
- If the trackStore modifications are done properly the selection of a track should be handled in there.
- May need to have the ability to create a new track for a specific camera instead of a generic new track.  This may need some more thought as to implementation.

### PR DESCRIPTION
**Pull request by BryonLewis**
Restructured the PRs for multicam into different feature branches.  Renamed a branch and modified some stuff to make it a bit more clear what is happening.

Closes #1094.

### CODE REVIEW

**Comment by BryonLewis** [comment]
I need to update the resizing of the windows based on the control size modification.  Currently it is using a number from a frame behind.  You can see this by collapsing/expanding the timeline area.  If you collapse it and resize the window slightly you'll see it resize properly.  The resize Observer is not calculating the proper height it is using the height before the timeline panel collapses down.

**Comment by BryonLewis** [comment]
Through TrackStore development found another issue:
Reloading datasets after import broke the geoViewer references.  Not exactly sure but I think it's because it calls a re-initialization of the cameras on top of existing values.  It seems throwing a clearing function into the useMediaController where all of the base references are reset to their default values fixes this.  I would like to know a bit more about what is going on and how it fixes it.

**Comment by BryonLewis** [comment]
List of changes since the merge:

- Modified using symbols back to using strings.  This has to do with typescript support for strings vs symbols and that we should have to update to a later version of TS for the symbol support for `Record<symbol, data>`
- When doing a reload on a page by either importing new annotation data or when a piplined had finished the system would reload properly but the annotations wouldn't show up.  I believe this had to do with creating a new 'symbol' with the exact same name as the one before.  The `clear()` function which resets the geoViewer references and symbols seemed to fix this issue when the reload happened.  The other option would be to prevent reloading of the layers when we reload all of the annotation data.  This may be a bit difficult because reloading the annotation data means that the camera information is reloaded again.

File: client/dive-common/components/Viewer.vue [hunk 1]
```diff
@@ -43,18 +43,28 @@ import { cloneDeep } from 'lodash';
...
 
+export interface ImageDataItem {
+  url: string;
+  filename: string;
+}
+
+function loadImageFunc(imageDataItem: ImageDataItem, img: HTMLImageElement) {
+  // eslint-disable-next-line no-param-reassign
+  img.src = imageDataItem.url;
+}
```
Comment (subdavis) [inline client/dive-common/components/Viewer.vue:54]: This is identical to the default prop value. Why is it specified here?

File: client/dive-common/components/Viewer.vue [hunk 2]
```diff
@@ -77,27 +87,18 @@ export default defineComponent({
     const loadError = ref('');
     const baseMulticamDatasetId = ref(null as string | null);
     const datasetId = toRef(props, 'id');
-    const multiCamList: Ref<string[]> = ref([]);
-    const defaultCamera = ref('');
-    const currentCamera = ref('');
+    const multiCamList: Ref<string[]> = ref(['default']);
+    const defaultCamera = ref('default');
+    const selectedCamera = ref('default');
```
Comment (subdavis) [inline client/dive-common/components/Viewer.vue:92]: I think we need to do with with something other than strings.  Using the reserved keyword "default" to name a camera causes the viewer to break.

File: client/dive-common/components/Viewer.vue [hunk 3]
```diff
@@ -77,27 +87,18 @@ export default defineComponent({
...
-    });
+    const { aggregateController, onResize, clear: mediaControllerClear } = useMediaController();
     const { time, updateTime, initialize: initTime } = useTimeObserver();
-    const imageData = ref([] as FrameImage[]);
+    const imageData = ref({ default: [] } as Record<string, FrameImage[]>);
     const datasetType: Ref<DatasetType> = ref('image-sequence');
     const datasetName = ref('');
     const saveInProgress = ref(false);
-    const videoUrl = ref(undefined as undefined | string);
+    const videoUrl = ref({ default: null } as Record<string, null | string>);
```
Comment (subdavis) [inline client/dive-common/components/Viewer.vue:101]: It's subtle, but there are two empty conditions here when there should only be one.

Either the key is missing, or the key is set to null. So you have to compare against both null and undefined. It would be more strict to choose a single empty condition, which is that if no videourl exists for a camera, its key is not in the map at all, so you only have to compare against undefined.

File: client/dive-common/components/Viewer.vue [hunk 4]
```diff
@@ -343,28 +349,54 @@ export default defineComponent({
...
+            const camTracks = Object.values(camTrackData);
+            progress.total = camTracks.length;
+            for (let j = 0; j < camTracks.length; j += 1) {
+              if (j % 4000 === 0) {
+              /* Every N tracks, yeild some cycles for other scheduled tasks */
+                progress.progress = j;
+                // eslint-disable-next-line no-await-in-loop
+                await new Promise((resolve) => window.setTimeout(resolve, 500));
+              }
+              insertTrack(Track.fromJSON(camTracks[j]), { imported: true, cameraName: camera });
```
Comment (subdavis) [inline client/dive-common/components/Viewer.vue:389]: We should avoid this logic duplication.

We can handle this by bumping the metadata schema version number and doing a meta.json migration when v1 is loaded so that Viewer.vue doesn't have to deal with this case.

In web, we can do a migration in mongo.

File: client/dive-common/use/useModeManager.ts [hunk 5]
```diff
@@ -224,7 +228,10 @@ export default function useModeManager({
 
   function handleUpdateRectBounds(frameNum: number, flickNum: number, bounds: RectBounds) {
     if (selectedTrackId.value !== null) {
-      const track = trackMap.get(selectedTrackId.value);
+      let track = trackMap.get(selectedTrackId.value);
+      if (selectedCamera.value !== 'default') {
+        track = camTrackMap[selectedCamera.value].get(selectedTrackId.value);
+      }
```
Comment (subdavis) [inline client/dive-common/use/useModeManager.ts:234]: I don't like having consumers responsible for all this logic, it's just more for bugs. 
 Possibly not part of this PR, but i think we need to 

1. Get rid of `trackMap` and replace it with `camTrackMap`
2. Replace everywhere that access the trackMap with a `getTrack(map, id, camera)` call that properly throws an error if either the camera or id don't exist.

File: client/dive-common/use/useModeManager.ts [hunk 6]
```diff
@@ -270,14 +277,20 @@ export default function useModeManager({
...
       if (track) {
         // newDetectionMode is true if there's no keyframe on frameNum
         const { features, interpolate } = track.canInterpolate(frameNum);
         const [real] = features;
 
         // Give each recipe the opportunity to make changes
         recipes.forEach((recipe) => {
+          if (!track) {
+            return;
+          }
```
Comment (subdavis) [inline client/dive-common/use/useModeManager.ts:293]: What is this doing?  We're inside an `if (track)` block already.

FWIW, that `if (track)` is also probably wrong because it implies that the else case is a valid execution path when it should probably be a runtime error.

File: client/dive-common/use/useModeManager.ts [hunk 7]
```diff
@@ -29,23 +29,27 @@ interface SetAnnotationStateArgs {
...
   selectNextTrack,
   addTrack,
   removeTrack,
 }: {
   selectedTrackId: Ref<TrackId | null>;
+  selectedCamera: Ref<string>;
   editingTrack: Ref<boolean>;
   trackMap: Map<TrackId, Track>;
-  mediaController: Ref<MediaController>;
+  camTrackMap: Record<string, Map<TrackId, Track>>;
```
Comment (subdavis) [inline client/dive-common/use/useModeManager.ts:47]: I don't think we should have `trackMap` anymore.

File: client/dive-common/components/Viewer.vue [hunk 8]
```diff
@@ -381,12 +414,30 @@ export default defineComponent({
...
+        onResize();
       }
-      const newId = `${baseMulticamDatasetId.value}/${camera}`;
-      ctx.emit('update:id', newId);
+    }
+    const observer = new ResizeObserver(handleResize);
+    watch(controlsRef, (previous) => {
+      if (previous) observer.unobserve(previous.$el);
+      if (controlsRef.value) observer.observe(controlsRef.value.$el);
+    });
```
Comment (subdavis) [inline client/dive-common/components/Viewer.vue:430]: Can you add a comment about what causes this watch to fire?  I think it's shadowing the value of the wrapping `v-if="progress.loaded"`, and I'm fine with doing it like this, I just want to make sure that's what's happening.

**Review by subdavis — CHANGES_REQUESTED** [review]
I don't know if some of this is supposed to work, so I'm just noting it here:

- [ ] Continuous mode seems broken.  When you start in continuous, then move into another camera, the "empty" annotation that gets created isn't removed, so if you're doing continuous, you end up with a bunch of detections with no geometry.
- [ ] Pressing `r` to recenter is broken

I ignored the trackstore.

File: client/dive-common/components/Viewer.vue (hunk 1 above)
Comment (BryonLewis) [reply client/dive-common/components/Viewer.vue:54]: I believe it's a remnant of the refactor, I'll remove it.

File: client/dive-common/use/useModeManager.ts (hunk 5 above)
Comment (BryonLewis) [reply client/dive-common/use/useModeManager.ts:234]: I've done this in the trackStore modification branch.  There really is no longer a trackMap, it's all `camMap` used in conjunction with different `getTrack` functions.  Except it's a little more complicated than a single `getTrack` function, because `getTrack(map, id, camera)` doesn't cover all needs, sometimes you just need the first trackId that's available and don't care what camera is selected, other times you need a conglomerate of all tracks across all cameras to display (in the case of the event Viewer for a selected track).  Finally you need a list of all available tracks to iterate over (For setting confidence pairs on all tracks regardless of camera, or setting track wide attributes)

Comment (subdavis) [reply client/dive-common/use/useModeManager.ts:234]: Got it, I remember talking about this now.  If we need some new functions or a more complicated `getTrack` that makes sense. I just don't want each function that needs to get a track to have to implement its own error handling / null checking.

File: client/dive-common/components/Viewer.vue (hunk 4 above)
Comment (subdavis) [reply client/dive-common/components/Viewer.vue:389]: Not important to resolve in this PR, just wanted to comment.

**Comment by subdavis** [comment]
Maybe do the easy stuff and let's resolve the harder stuff later?

* `'default'` reserved keyword is an issue, but not critical.
* `trackMap` is taken care of in your next PR
* Can think about doing a schema migration for representing single and multi-cam data in a consistent way after the other stuff works.

File: client/dive-common/use/useModeManager.ts (hunk 6 above)
Comment (BryonLewis) [reply client/dive-common/use/useModeManager.ts:293]: it's a typescript oddity that it believe that track can be undefined even though the forEach is inside of an `if (track)`.  I'll see what I can do about it.

Comment (waxlamp) [reply client/dive-common/use/useModeManager.ts:293]: > it's a typescript oddity

This would go beyond "oddity"--either there's some weird dynamics in this code that we're missing (perhaps due to lexical scoping of the `forEach()` callback?), or there's a typescript bug. What error do you get when you omit the if block? You can also use this technique (https://stackoverflow.com/a/66124742/1886928) to force the compiler to tell you what it thinks the type of `track` is at that point in the code.

Comment (BryonLewis) [reply client/dive-common/use/useModeManager.ts:293]: It believes that within the forEach that track can be `Track | undefined` and doesn't want to use it as an argument to `recipe.update`
image

Comment (waxlamp) [reply client/dive-common/use/useModeManager.ts:293]: And how about outside of the forEach callback (3 lines up, on the `track.canInterpolate()` invocation)?

Comment (BryonLewis) [reply client/dive-common/use/useModeManager.ts:293]: It's fine there and is known as a track.  I've updated my last comment and it might have been too late.  But there is an issue exactly like this and has a simplistic demo showing the problem.  I think I've seen this before as well.
https://github.com/microsoft/TypeScript/issues/29916

Comment (subdavis) [reply client/dive-common/use/useModeManager.ts:293]: Thanks for looking into this.

I know this is nitpicky, but it happens from time to time.  When typescript misbehaves, I think it would be better to put in a `// @ts-ignore` than to insert code that handles what we know to be an invalid execution path.  That `return` should be unreachable, and I wouldn't blame a reasonable developer for seeing that in the future and ripping it out because it looks like it shouldn't be there.

Comment (subdavis) [reply client/dive-common/use/useModeManager.ts:293]: And because you don't want to disable type checks for that whole line, you can probably add something like this:

``` ts
// @ts-ignore
const _track: Track = track;
const changes = recipe.update(eventType, frameNum, _track, [data], key);
```

Comment (BryonLewis) [reply client/dive-common/use/useModeManager.ts:293]: Was implementing and did a quick search for the eslint-disable command and found this:
image
apparently my mind isn't deteriorating as quickly as I thought, we have run into this before.

**Comment by BryonLewis** [comment]
I've switch to `'singleCam'` for the identifier for a singleCamera.  Right now it will conflict if there is a multiCam with the name of `'singleCam'`.  It will be easier to resolve after I re-base my trackstore PR.  There all `trackMaps` are within a `camMap` meaning I can default to camera easier.
I think in the next PR I can also modify the Loading mechanism to remove the duplicate logic because the change to have `camMap: Map<string,<Map<trackId, Track>>>` be the holder of all tracks instead of this ridiculous thing I did temporarily here to get something to work.

File: client/dive-common/use/useModeManager.ts (hunk 6 above)
Comment (waxlamp) [reply client/dive-common/use/useModeManager.ts]: You can also use a type assertion here ("casting" with the `as` operator) to avoid having to disable the ts-lint rules.

**Review by subdavis — APPROVED** [review]
wooo
